### PR TITLE
[Client] [Tooling] refactor: CLI

### PR DESCRIPTION
**Pull request by bryanchriswhite**
### Summary generated by Reviewpad on 13 Jun 23 12:53 UTC
This pull request includes various updates across the codebase, including changes to CLI flags, import statements, and default values. It also involves refactoring code to use shared helper functions, removing unused code, and improving error handling. Some variables have been renamed to be more descriptive. The specific changes include modifications to files such as `validator.go`, `cluster-manager.yaml`, `cli-client.yaml`, `utils.go`, and `runtime_defaults.go`. For example, the `RPC_HOST` variable has been replaced with `POCKET_REMOTE_CLI_URL`. A new `helper` package has been added, and a new `flags` package contains several CLI flags. Code has also been updated to use `flags.RemoteCLIURL` instead of `remoteCLIURL`. Some constants have been renamed to include `Hostname`, and there are updates to endpoint hostnames for some validators.

## Issue

Dependants:
- #730

## Type of change

Please mark the relevant option(s):

- [ ] New feature, functionality or library
- [ ] Bug fix
- [x] Code health or cleanup
- [ ] Major breaking change
- [ ] Documentation
- [ ] Other 

## List of changes

- Exported `rootCmd` to support cross-package usage (i.e. subcommands w/ own pkgs)
- Refactored common CLI code
  - Moved & refactored `PersistentPreRunE()` helper
  - Moved & exported `busCLICtxKey`
  - Exported `GetValueFromCLIContext()` and `SetValueInCLIContext()`
  - Moved `setupAndStartP2PModule`
  - Moved `setupCurrentHeightProvider`
  - Moved `setupPeerstoreProvider`
- Refactored CLI flags to own package for cross-package use
- Replaced RPC_HOST with POCKET_REMOTE_CLI_URL where appropriate
- Added `remote-cli-url` flag to cluster manager & refactor
- Added support for overriding of `remote-cli-url` flag with env var
- Append "Hostname" to validator endpoint hostname constants
- Promoted string literal to `RandomValidatorEndpointK8SHostname` constant
- Improved e2e tests error messaging
- Simplify e2e debug command calls
- Improve error handling in client CLI

## Testing

- [ ] `make develop_test`; if any code changes were made
- [ ] `make test_e2e` on [k8s LocalNet](https://github.com/pokt-network/pocket/blob/main/build/localnet/README.md); if any code changes were made
- [ ] `e2e-devnet-test` passes tests on [DevNet](https://pocketnetwork.notion.site/How-to-DevNet-ff1598f27efe44c09f34e2aa0051f0dd); if any code was changed
- [x] [Docker Compose LocalNet](https://github.com/pokt-network/pocket/blob/main/docs/development/README.md); if any major functionality was changed or introduced
- [x] [k8s LocalNet](https://github.com/pokt-network/pocket/blob/main/build/localnet/README.md); if any infrastructure or configuration changes were made

## Required Checklist

- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added, or updated, [`godoc` format comments](https://go.dev/blog/godoc) on touched members (see: [tip.golang.org/doc/comment](https://tip.golang.org/doc/comment))
- [x] I have tested my changes using the available tooling
- [x] I have updated the corresponding CHANGELOG

### If Applicable Checklist

- [ ] I have updated the corresponding README(s); local and/or global
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added, or updated, [mermaid.js](https://mermaid-js.github.io) diagrams in the corresponding README(s)
- [ ] I have added, or updated, documentation and [mermaid.js](https://mermaid-js.github.io) diagrams in `shared/docs/*` if I updated `shared/*`README(s)

### CODE REVIEW

**Review by dylanlott — COMMENTED** [review]
Looks good, I like the flags package organization. Let's see it pass in DevNet and it should be good to go.

File: app/client/cli/helpers/setup.go [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+	"github.com/pokt-network/pocket/p2p"
+	rpc2 "github.com/pokt-network/pocket/p2p/providers/current_height_provider/rpc"
+	"github.com/pokt-network/pocket/p2p/providers/peerstore_provider/rpc"
+	"github.com/pokt-network/pocket/runtime"
+	"github.com/pokt-network/pocket/runtime/defaults"
+	"github.com/pokt-network/pocket/shared/modules"
+)
+
+// P2PDependenciesPreRunE initializes peerstore & current height providers, and a
+// p2p module with consumes them. Everything is registered to the bus.
```
[the remark this answers is not in the document]
Comment (bryanchriswhite) [reply app/client/cli/helpers/setup.go:19]: Yes, nice catch - thanks! 🙌

File: app/client/cli/flags/flags.go [hunk 2]
```diff
@@ -0,0 +1,13 @@
+package flags
```
[the remark this answers is not in the document]
Comment (Olshansk) [reply app/client/cli/flags/flags.go:1]: ❤️

File: app/client/cli/flags/flags.go [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+
+var (
+	RemoteCLIURL string
```
Comment (Olshansk) [inline app/client/cli/flags/flags.go:4]: Do these need to be exposed outside of the `flags` package?

File: app/client/cli/flags/flags.go [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+
+	DataDir string
+
+	ConfigPath string
+
+	NonInteractive bool
+
+	Verbose bool
```
Comment (Olshansk) [inline app/client/cli/flags/flags.go:12]: Can you add a very short comment for each of them?

File: app/client/cli/helpers/common.go [hunk 5]
```diff
@@ -0,0 +1,14 @@
+package helpers
+
+import (
+	"github.com/pokt-network/pocket/runtime"
+	"github.com/pokt-network/pocket/shared/modules"
+)
+
+var (
+	GenesisPath = runtime.GetEnv("GENESIS_PATH", "build/config/genesis.json")
```
Comment (Olshansk) [inline app/client/cli/helpers/common.go:9]: Reevaluate the visibility of the vars (from a package scope perspective)

File: app/client/cli/helpers/common.go [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+	RpcHost     string
```
Comment (Olshansk) [inline app/client/cli/helpers/common.go:10]: Can this be overriden by `RemoteCLIURL`? 

Asking to:
1. Understand
2. Modify / consolidate as needed

File: app/client/cli/helpers/common.go (hunk 5 above)
Comment (Olshansk) [inline app/client/cli/helpers/common.go:9]: Add a TODO to:
1. Also accept reading this from datadir
2. Accept this is a flag

File: app/client/cli/helpers/context.go [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+// NOTE: this is required by the linter, otherwise a simple string constant would have been enough
+type cliContextKey string
+
+func SetValueInCLIContext(cmd *cobra.Command, key cliContextKey, value any) {
+	cmd.SetContext(context.WithValue(cmd.Context(), key, value))
+}
+
+func GetValueFromCLIContext[T any](cmd *cobra.Command, key cliContextKey) (T, bool) {
+	value, ok := cmd.Context().Value(key).(T)
+	return value, ok
```
Comment (Olshansk) [inline app/client/cli/helpers/context.go:20]:
```suggestion
	return cmd.Context().Value(key).(T)
```

**Review by Olshansk — CHANGES_REQUESTED** [review]

File: app/client/cli/helpers/context.go (hunk 7 above)
Comment (bryanchriswhite) [reply app/client/cli/helpers/context.go:20]: I don't think you can return the multiple return args form of a type assertion directly. Same goes for retrieving a value from a map.

File: app/client/cli/flags/flags.go (hunk 3 above)
Comment (bryanchriswhite) [reply app/client/cli/flags/flags.go:4]: Yes, all flag vars must be exported in order to be used as they have been moved into their own module.

File: app/client/cli/cmd.go [hunk 8]
```diff
@@ -2,40 +2,43 @@ package cli
...
-	verbose        bool
-	cfg            *configs.Config
+	cfg *configs.Config
 )
 
 func init() {
-	rootCmd.PersistentFlags().StringVar(&remoteCLIURL, "remote_cli_url", defaults.DefaultRemoteCLIURL, "takes a remote endpoint in the form of <protocol>://<host> (uses RPC Port)")
-	rootCmd.PersistentFlags().BoolVar(&nonInteractive, "non_interactive", false, "if true skips the interactive prompts wherever possible (useful for scripting & automation)")
+	rootCmd.PersistentFlags().StringVar(&flags.RemoteCLIURL, "remote_cli_url", defaults.DefaultRemoteCLIURL, "takes a remote endpoint in the form of <protocol>://<host>:<port> (uses RPC Port)")
+	// ensure that this flag can be overidden by the respective viper-conventional environment variable (i.e. `POCKET_REMOTE_CLI_URL`)
```
Comment (okdas) [inline app/client/cli/cmd.go:25]: Have you checked if it environment variable works? I love that you bind a viper configuration instead of looking up the environment variable!

Comment (bryanchriswhite) [reply app/client/cli/cmd.go:25]: I believe so. The CLI client and cluster manager containers are using `POCKET_REMOTE_CLI_URL` in this branch and both k8s and docker-compose localnets seem to be working as expected. I even used a fresh kind cluster for the k8s localnet.

Comment (bryanchriswhite) [reply app/client/cli/cmd.go:25]: :thinking: hmm... it doesn't seem to be working in the context of `make localnet_shell` though when it seems like it should, perhaps I've missed something. Taking another look :eyes:

Comment (bryanchriswhite) [reply app/client/cli/cmd.go]: @okdas thanks for making me double check this. :sweat_smile: 

I did a bit of investigating and I'm noticing that the way `BindPFlag()` works is not quite what I expected. It seems to "bind the respective viper key" in the sense that the final result will be reflected in the value retrieved from viper (as opposed to updating the bound flag variable). 0b22fbd adds a line just after the call to `ParseConfig()` which updates the flag var to this final value. My thinking is that it's better to keep the flag var than to replace all references to it with `viper.GetString("remote_cli_url")` or something similar.

File: e2e/tests/validator.go [hunk 9]
```diff
@@ -18,7 +18,19 @@ var (
 )
 
 func init() {
-	rpcURL = fmt.Sprintf("http://%s:%s", runtime.GetEnv("RPC_HOST", "pocket-validators"), defaults.DefaultRPCPort)
+	// set the rpcURL based on the environment while still supporting overriding of the RPC_HOST.
+	var rpcHost string
+	// TECHDEBT: if we intend to support running the e2e tests in both tilt/k8s and
```
Comment (Olshansk) [inline e2e/tests/validator.go:23]: I think E2E tests are only intended for tilt/k8s - not docker compose

File: e2e/tests/validator.go [hunk 10]
```diff
@@ -18,7 +18,19 @@ var (
...
+	// docker compose, then we need to understand why the `KUBERNETES_SERVICE_HOST`
+	// isn't set even when running in k8s. See: `test_go('e2e-tests', ...` in the
+	// Tiltfile.
+	// if runtime.IsProcessRunningInsideKubernetes() {
```
Comment (Olshansk) [inline e2e/tests/validator.go:27]: Is this a WIP to be checked with @okdas in this PR?

File: build/localnet/cluster-manager/main.go [hunk 11]
```diff
@@ -31,13 +35,45 @@ var (
...
+	// the default, which is set to the URL of the first k8s validator's RPC
+	// endpoint.
+	clusterManagerCmd.PersistentFlags().StringVar(
+		&flags.RemoteCLIURL,
+		"remote_cli_url",
+		defaults.Validator1EndpointK8SHostname,
+		"takes a remote endpoint in the form of <protocol>://<host>:<port> (uses RPC Port)",
+	)
+
+	// ensure that the env var can override the flag
```
Comment (Olshansk) [inline build/localnet/cluster-manager/main.go:64]: I know we've had a lot of issues with viper and flag binding in the past.

I don't have much feedback on this as I'd have to dive in really deep, but have to be honest that I have lost context on where/how this variable binding is working.

Anything I should focus on or provide feedback in particular? If not, LGTM

**Review by Olshansk — APPROVED** [review]
@bryanchriswhite Aside from the failing E2E tests, this LGTM and I don't have much feedback to provide.

I know you still have to resolve some E2E failures but doing a "soft approval" in the meantime so we can get this over the finish line.

File: app/client/cli/helpers/setup.go [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+	"github.com/pokt-network/pocket/logger"
+	"github.com/pokt-network/pocket/p2p"
+	rpcCHP "github.com/pokt-network/pocket/p2p/providers/current_height_provider/rpc"
+	rpcPSP "github.com/pokt-network/pocket/p2p/providers/peerstore_provider/rpc"
+	"github.com/pokt-network/pocket/runtime"
+	"github.com/pokt-network/pocket/shared/modules"
+)
+
+// P2PDependenciesPreRunE initializes peerstore & current height providers, and a
+// p2p module with consumes them. Everything is registered to the bus.
```
Comment (bryanchriswhite) [inline app/client/cli/helpers/setup.go:16]:
```suggestion
// p2p module which consumes them. Everything is registered to the bus.
```

File: app/client/cli/cmd.go [hunk 13]
```diff
@@ -43,6 +48,9 @@ var rootCmd = &cobra.Command{
 	PersistentPreRunE: func(cmd *cobra.Command, args []string) error {
 		// by this time, the config path should be set
 		cfg = configs.ParseConfig(flags.ConfigPath)
+
+		// set final `remote_cli_url` value; order of precedence: flag > env var > config > default
```
[the remark this answers is not in the document]
Comment (bryanchriswhite) [reply app/client/cli/cmd.go:52]: It would probably be a good idea, for consistency across the CLI UX, to do that, yea.

File: app/client/cli/flags/flags.go [hunk 14]
```diff
@@ -1,13 +1,24 @@
 package flags
 
 var (
+	// RemoveCLIURL is the URL of the remote RPC node which the CLI will interact with.
+	// Formatted as <protocol>://<host>:<port> (uses RPC Port).
+	// (see: --help the root command for more info).
```
[the remark this answers is not in the document]
Comment (bryanchriswhite) [reply app/client/cli/flags/flags.go:6]: Fair enough, I was imagining browsing the godoc, perhaps a package-level comment would be more appropriate.

File: e2e/tests/validator.go (hunk 10 above)
Comment (bryanchriswhite) [reply e2e/tests/validator.go:27]: If we don't care about running e2e tests in docker compose, I would just clean this up.

File: build/localnet/cluster-manager/main.go (hunk 11 above)
Comment (bryanchriswhite) [reply build/localnet/cluster-manager/main.go:64]: What I think I've learned regarding bindings: cobra takes the default values and handles updating bound flag variables/pointers based on any actual flag values which are present during invocation (i.e. `--remote_cli_url someValue`). Separately from that, viper can track a key of the same name, the value of which (independent of the flag vars/ptrs) will default to its respective config value (if present/applicable) but can be overridden by a respective env var and/or the flag value.

TLDR; the viper/cobra API dynamic is such that:
1. Viper key/values are separate (in memory) from viper flags.
2. Viper key/values can be overridden by cobra flags but not the other way around.
3. For consistency we should either
  - Ensure we're updating flag vars (i.e. bound to a cobra flag) with their respective viper overrides
  - Stop using flag vars (e.g. do `#PersistentFlags.String(...)` instead of `#PersistentFlags.StringVar(&flagVar, ...)` and treat viper as the single source of truth (e.g. `viper.GetString(...)`)

File: app/client/cli/flags/flags.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (bryanchriswhite) [reply app/client/cli/flags/flags.go:6]: (Adding this to my list of TODOs in a downstream PR to avoid another round with CI)

**Comment by bryanchriswhite** [comment]
:rotating_light:  (please don't delete this branch yet - I will do so once upstream branches have been rebased or merged with main) :rotating_light:
